In [46]:
from IPython.core.display import HTML
HTML('<a href="https://colab.research.google.com/github/yingchen0817/CSC448/blob/main/Assignment3.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>')

In [69]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn import linear_model

# Part 1. Loading the dataset

In [48]:
# Using pandas load the dataset (load remotely, not locally)
slime = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")


In [49]:
# read in the data
slime

Temperature °C  Mols KCL     Size nm^3
0               469       647  6.244743e+05
1               403       694  5.779610e+05
2               302       975  6.196847e+05
3               779       916  1.460449e+06
4               901        18  4.325726e+04
..              ...       ...           ...
995             894       847  1.545661e+06
996             327       982  6.737041e+05
997             791       213  3.477543e+05
998             769       553  8.684794e+05
999             919       452  8.476413e+05

[1000 rows x 3 columns]

In [50]:
# Output the first 15 rows of the data
slime.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [51]:
# Display a summary of the table information (number of datapoints, etc.)
slime.info()
slime.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

In [52]:
# number of datapoint
slime.size

3000

# Part 2. Splitting the dataset

In [53]:
# Take the pandas dataset and split it into our features (x) and label (y)
features = slime[['Temperature °C', 'Mols KCL']]
label = slime['Size nm^3']

In [54]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9)

In [55]:
# display
x_train, x_test, y_train, y_test

(     Temperature °C  Mols KCL
 208             484       863
 998             769       553
 62               61       283
 963             754        71
 449              49       227
 ..              ...       ...
 25              899         5
 923             381       493
 438             366       627
 171             521       161
 324             703       774
 
 [900 rows x 2 columns],
      Temperature °C  Mols KCL
 530             858       835
 260             106       457
 183             689       500
 138             209       299
 18              905        76
 ..              ...       ...
 298             985       786
 268             882       900
 966             152       380
 81              291       709
 184             980       277
 
 [100 rows x 2 columns],
 208    8.624711e+05
 998    8.684794e+05
 62     3.754626e+04
 963    1.162600e+05
 449    2.430626e+04
            ...     
 25     1.977871e+04
 923    3.871823e+05
 438    4.745883e+05
 171    1.747

# Part 3. Perform a Linear Regression


In [70]:
# Use sklearn to train a model on the training set
LinReg = LinearRegression().fit(x_train, y_train) # fit
LinReg.score(x_test, y_test) # score

0.8665282968262333

In [81]:
# Create a sample datapoint and predict the output of that sample with the trained model
LinReg.predict(np.array([[1,2]]))

/usr/local/Cellar/jupyterlab/3.2.9/libexec/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([-405778.03512925])

In [74]:
LinReg.coef_

array([ 871.5446807 , 1021.85246031])

In [62]:
LinReg.intercept_ 

-408693.28473056445

In [63]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
# Extract the coefficents and intercept from the model and write an equation for your h(x)

### Line of the best fit will be (LinReg.intercept + LinReg.coef) which will not be 100% accuracute.  The equation was created with the coefficents a, b and the intercept that fit h(x).

### y = -408693.28473056445 + 871.5446807 + 1021.85246031
### y = -406799.88759 whihc is predict (h(x))



# Part 4. Use Cross Validation

In [68]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cross_val_score(LinReg, x_test, y_test, cv=5)

array([0.89338431, 0.86620625, 0.82118291, 0.85277056, 0.83900783])

In [ ]:
# Report on their finding and their significance

##### this values trying to get close to the true value of LinReg from 5 cv run

# Part 5. Using Polynomial Regression

In [75]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
PolyReg = PolynomialFeatures(degree = 2)
PolyReg_train = PolyReg.fit_transform(x_train)
PolyReg_test = PolyReg.fit_transform(x_test)

BayReg = linear_model.BayesianRidge().fit(PolyReg_train, y_train)
BayReg.score(PolyReg_test, y_test)

1.0

In [76]:
BayReg.coef_

array([ 0.00000000e+00,  1.20000000e+01, -1.16474838e-07, -1.10938351e-11,
        2.00000000e+00,  2.85714287e-02])

In [77]:
BayReg.intercept_

1.4421762898564339e-05

In [82]:

BayReg.predict(np.array([[1, 2, 3, 4, 5, 6]]))

array([34.17144266])

In [83]:
# Report on the metrics and output the resultant equation as you did in Part 3.

#### y = 1.4421762898564339 * 10^-5 + 0.00000000 * 1 + 1.20000000 * 10^1 * a + -1.16474838 * 10^-7 * b + -1.10938351 * 10^-11 * (a^2) + 2.00000000 * ab + 2.85714287 * 10^-2 * (b^2)
#### we can see the above value (1, a, b, a^2, ab, b^2 ) is from the coefficients arrays

#### we are using polynomial features is because this is a polynomial regression, we then can convert 2D matrix into the regular matrix and calculate it. we can easily fit the data into the model with accurate value.